### DISCUSSION TITLE
WhenAnyValue with null propogation, how to init and monitor for other values?

### DISCUSSION THREAD

**Discussion by sinatrocious** [reactiveui/ReactiveUI#3770] [Q&A] [answered]
There is a proposed [workaround](https://www.reactiveui.net/docs/handbook/when-any#null-propogation-inside-whenanyvalue) for null propagation. However it doesn't trigger (at all) if expression has null value in the chain. Which leads to 2 problems:

- it doesn't provide initial value (can be worked around with `StartWith()` >< )
- it will not be able to monitor for changes of other observables.

I made a simple test to demonstrate the issue.

<details>
  <summary>Show code</summary>

```c#
public class Program
{
    static void Main(string[] args)
    {
        var parent = new Parent();

        parent.WhenAnyValue(o => o.Child, o => o.Trigger, o => o.Child!.Property)
            .Subscribe(o => Console.WriteLine($"A: {o.Item1}"));

        parent.WhenAnyValue(o => o.Child, o => o.Trigger, o => o.Child!.Property)
            .StartWith((null, false, null))
            .Subscribe(o => Console.WriteLine($"B: {o.Item1}"));

        parent.WhenAnyValue(o => o.Trigger)
            .CombineLatest(parent.WhenAnyValue(o => o.Child, o => o.Child!.Property))
            .Subscribe(o => Console.WriteLine($"C: {o.Second.Item1}"));

        parent.WhenAnyValue(o => o.Child, o => o.Trigger)
            .Subscribe(o => Console.WriteLine($"D: {o.Item1}"));

        Console.WriteLine("Trigger");

        parent.Trigger = true;

        Console.ReadKey();
    }
}

public class Parent : ReactiveObject
{
    [Reactive]
    public Child? Child { get; set; }

    [Reactive]
    public bool Trigger { get; set; }

    public override string ToString() => $"{Trigger} {Child}";
}

public class Child : ReactiveObject
{
    [Reactive]
    public string? Property { get; set; }

    public override string ToString() => $"Child {Property}";
}
```
</details>

I have to monitor for `Trigger` and `Child.Property` properites.

Run the test and the output will be:

> B:
> D:
> Trigger
> D:

In other words:
- `A` never works.
- `B` with `StartWith` workaround only able to init value, but then doesn't work.
- `C` (attempt to use `CombineLatest`) never works either
- only `D` (doesn't have null propagation) **works**, but it's not able to catch `Child.Property` changes.

Question
---

How to monitor for `Trigger` and `Child.Property` (where `Child` may always be `null`)?

**Comment by sinatrocious** [comment, marked as answer]
Found a workaround!

For null propagated property it's absolutely necessary to create an **extra property** to monitor only for value changes of that property and then use that extra property instead of null-forgiving operator. Then `WhenAnyValue` works for all scenarios.

To fix code from first post:

1. Add extra property to `parent` (it can be anywhere, but makes the most sense to be there):

```c#
[ObservableAsProperty]
public string? ChildProperty { get; }
```

2. Use 2 subscriptions:

```c#
// this subscription only monitors for child property value
// it will stop working as any WhenAnyValue with null-forgiving operator when parent in expression is null
// but we don't care, as we only need correct value when parent is not null
parent.WhenAnyValue(o => o.Child, o => o.Child!.Property, (child, property) => property)
    .ToPropertyEx(parent, o => o.ChildProperty);

// this subscription is bullet proof and will correctly rise on any change!
parent.WhenAnyValue(o => o.Child, o => o.Trigger, o => o.ChildProperty)
    .Subscribe(o => Console.WriteLine($"E: {o.Item1}"));
```

3. Test it, it should trigger after any change:

```c#
parent.Trigger = !parent.Trigger; // E:

parent.Child = new(); // E: Child

parent.Child.Property = "123"; // E: Child 123

parent.Child = null; // E:
```
